File: packages/sdk-ts/src/kernel/bundle-handle.ts
/**
 * BundleHandle — source-agnostic interface for reading + writing a
 * bundle's entries.
 *
 * 1:1 parity with python/dna/kernel/bundle_handle.py.
 *
 * Implementations:
 *   - `FilesystemBundleHandle` — wraps a real filesystem directory.
 *   - `DictBundleHandle` — in-memory, used by SQL/Postgres adapters
 *     (rehydrate bundle entries from `dna_bundle_entries` rows into
 *     a dict, hand to readers).
 *   - `PostgresBundleHandle` (Postgres adapter) — backed by
 *     `dna_bundle_entries` rows.
 *
 * Entry naming convention: a posix-style relative path inside the
 * bundle. Top-level entries are bare names (`"SKILL.md"`,
 * `"IDENTITY.md"`); nested entries use forward slashes
 * (`"scripts/run.py"`, `"references/spec.md"`).
 *
 * v1.0 async refactor: read/write methods return `Promise<...>`.
 * Async-everywhere is the contract — sync local-FS impls wrap their
 * sync IO in `Promise.resolve` (true `fs/promises` migration is a
 * v1.2 cleanup).
 */

import { access, readFile, readdir, stat, writeFile, mkdir } from "node:fs/promises";
import { join, dirname, basename, posix } from "node:path";

/** True iff the path exists. fs/promises doesn't ship a boolean
 *  `exists`; this is the canonical workaround. */
async function pathExists(p: string): Promise<boolean> {
  try { await access(p); return true; } catch { return false; }
}

export interface BundleHandle {
  /** Bundle directory name (used as default doc name when frontmatter
   *  omits `metadata.name`). */
  readonly name: string;

  /** True if the named entry (file or directory) exists in this bundle. */
  exists(entry: string): Promise<boolean>;

  /** Read entry content as text. Throws if absent. */
  readText(entry: string, encoding?: BufferEncoding): Promise<string>;

  /** Read entry content as bytes. Throws if absent. */
  readBytes(entry: string): Promise<Uint8Array>;

  /**
   * Yield entry names (relative to the bundle root).
   *
   * When `recursive=false` (default), only direct children are
   * yielded — both regular files and subdirectories.
   * When `recursive=true`, descend into subdirectories yielding only
   * regular files (no directory entries).
   */
  iterEntries(recursive?: boolean): Promise<string[]>;

  /** True if `entry` points at a regular file (not a directory). */
  isFile(entry: string): Promise<boolean>;

  /** Write text content. Read-only handles throw. */
  writeText(entry: string, content: string, encoding?: BufferEncoding): Promise<void>;

  /** Write bytes. Read-only handles throw. */
  writeBytes(entry: string, content: Uint8Array): Promise<void>;

  /**
   * Filesystem path when the handle wraps a real directory; null
   * otherwise. ESCAPE HATCH — prefer explicit read/write/iter methods.
   * Use only when an external library demands a real path (e.g.
   * `fs.cp`, third-party tooling).
   */
  readonly path: string | null;
}

// ---------------------------------------------------------------------------
// Filesystem implementation
// ---------------------------------------------------------------------------

export class FilesystemBundleHandle implements BundleHandle {
  constructor(private readonly _root: string) {}

  get name(): string { return basename(this._root); }
  get path(): string | null { return this._root; }

  async exists(entry: string): Promise<boolean> {
    return pathExists(join(this._root, entry));
  }

  async readText(entry: string, encoding: BufferEncoding = "utf-8"): Promise<string> {
    const p = join(this._root, entry);
    try {
      return await readFile(p, encoding);
    } catch (e) {
      if ((e as NodeJS.ErrnoException).code === "ENOENT") {
        const err = new Error(`Bundle entry not found: ${entry}`) as Error & { code: string };
        err.code = "ENOENT";
        throw err;
      }
      throw e;
    }
  }

  async readBytes(entry: string): Promise<Uint8Array> {
    const p = join(this._root, entry);
    try {
      return new Uint8Array(await readFile(p));
    } catch (e) {
      if ((e as NodeJS.ErrnoException).code === "ENOENT") {
        const err = new Error(`Bundle entry not found: ${entry}`) as Error & { code: string };
        err.code = "ENOENT";
        throw err;
      }
      throw e;
    }
  }

  async iterEntries(recursive: boolean = false): Promise<string[]> {
    if (!(await pathExists(this._root))) return [];
    const out: string[] = [];
    if (recursive) {
      const walk = async (dir: string, prefix: string): Promise<void> => {
        const entries = (await readdir(dir)).sort();
        for (const entry of entries) {
          const full = join(dir, entry);
          const rel = prefix ? posix.join(prefix, entry) : entry;
          const st = await stat(full);
          if (st.isDirectory()) await walk(full, rel);
          else out.push(rel);
        }
      };
      await walk(this._root, "");
    } else {
      for (const entry of (await readdir(this._root)).sort()) out.push(entry);
    }
    return out;
  }

  async isFile(entry: string): Promise<boolean> {
    const p = join(this._root, entry);
    try {
      return (await stat(p)).isFile();
    } catch {
      return false;
    }
  }

  async writeText(entry: string, content: string, encoding: BufferEncoding = "utf-8"): Promise<void> {
    const p = join(this._root, entry);
    await mkdir(dirname(p), { recursive: true });
    await writeFile(p, content, encoding);
  }

  async writeBytes(entry: string, content: Uint8Array): Promise<void> {
    const p = join(this._root, entry);
    await mkdir(dirname(p), { recursive: true });
    await writeFile(p, Buffer.from(content));
  }
}

// ---------------------------------------------------------------------------
// In-memory implementation (used by tests + SQL adapters)
// ---------------------------------------------------------------------------

export class DictBundleHandle implements BundleHandle {
  private _entries: Map<string, string | Uint8Array>;

  constructor(public readonly name: string, entries: Record<string, string | Uint8Array> = {}) {
    this._entries = new Map(Object.entries(entries));
  }

  get path(): string | null { return null; }

  async exists(entry: string): Promise<boolean> {
    if (this._entries.has(entry)) return true;
    // Treat parent paths as "directory entries" — true if any key
    // starts with `entry/`.
    const prefix = entry.endsWith("/") ? entry : `${entry}/`;
    for (const k of this._entries.keys()) if (k.startsWith(prefix)) return true;
    return false;
  }

  async readText(entry: string, _encoding: BufferEncoding = "utf-8"): Promise<string> {
    const v = this._entries.get(entry);
    if (v === undefined) {
      const err = new Error(`Bundle entry not found: ${entry}`) as Error & { code: string };
      err.code = "ENOENT";
      throw err;
    }
    return typeof v === "string" ? v : new TextDecoder().decode(v);
  }

  async readBytes(entry: string): Promise<Uint8Array> {
    const v = this._entries.get(entry);
    if (v === undefined) {
      const err = new Error(`Bundle entry not found: ${entry}`) as Error & { code: string };
      err.code = "ENOENT";
      throw err;
    }
    return typeof v === "string" ? new TextEncoder().encode(v) : v;
  }

  async iterEntries(recursive: boolean = false): Promise<string[]> {
    if (recursive) return Array.from(this._entries.keys()).sort();
    // Direct children only — extract first segment, deduplicate
    const directChildren = new Set<string>();
    for (const k of this._entries.keys()) {
      const slash = k.indexOf("/");
      directChildren.add(slash === -1 ? k : k.substring(0, slash));
    }
    return Array.from(directChildren).sort();
  }

  async isFile(entry: string): Promise<boolean> {
    return this._entries.has(entry);
  }

  async writeText(entry: string, content: string, _encoding: BufferEncoding = "utf-8"): Promise<void> {
    this._entries.set(entry, content);
  }

  async writeBytes(entry: string, content: Uint8Array): Promise<void> {
    this._entries.set(entry, content);
  }
}

File: packages/sdk-py/dna/kernel/bundle_handle.py
"""BundleHandle — source-agnostic view of a single bundle directory.

A bundle is a logical container holding a marker file (AGENTS.md, SOUL.md,
SAFETYPOLICY.md, program.md, ...) plus optional sidecar files (scripts/,
references/, IDENTITY.md, etc.). Readers and writers receive a BundleHandle
and operate through this interface instead of pathlib / os.path — so the
same reader works whether the bundle lives on the filesystem, in a Postgres
row group, or in S3 / GCS / etc.

Background: Phase 8 audit (docs/superpowers/plans/2026-04-24-phase-8-port-cleanliness.md)
found that ReaderPort.detect/read and WriterPort.write were typed `path: Path`,
forcing every backing store to materialise a real filesystem dir before
invoking readers — which Postgres / S3 cannot do. By switching to
`BundleHandle`, Postgres adapter (PR2) can implement `PostgresBundleHandle`
backed by an `dna_bundle_entries` table and reuse all existing readers.

Migration philosophy: existing readers get a backward-compat escape hatch
via ``handle.path: Path | None`` — when the handle wraps a real directory,
the property returns it; otherwise None. Code that genuinely needs Path
semantics (e.g. shutil.copy) can opt in explicitly. The expectation is
that this property goes away over time as more backends ship.
"""
from __future__ import annotations

from pathlib import Path
from typing import Iterator, Protocol, runtime_checkable


@runtime_checkable
class BundleHandle(Protocol):
    """Source-agnostic interface for reading + writing a bundle's entries.

    Implementations:
      - ``FilesystemBundleHandle`` (this module) — wraps ``pathlib.Path``.
      - ``DictBundleHandle`` (this module) — in-memory, used in tests.
      - ``DictBundleHandle`` is also how the SQL adapter serves bundles —
        hydrated from ``dna_bundle_entries`` rows.

    Entry naming convention: a posix-style relative path inside the bundle.
    Top-level entries are bare names (``"SAFETYPOLICY.md"``,
    ``"IDENTITY.md"``); nested entries use forward slashes
    (``"scripts/run.py"``, ``"references/spec.md"``).
    """

    @property
    def name(self) -> str:
        """Bundle directory name (e.g. ``'talent-screener'``,
        ``'pii-ml-filter'``). Used by readers as a default doc name when
        the marker frontmatter omits ``metadata.name``.
        """
        ...

    def exists(self, entry: str) -> bool:
        """True if the named entry (file or directory) exists in this bundle."""
        ...

    def read_text(self, entry: str, encoding: str = "utf-8") -> str:
        """Read entry content as text. Raises ``FileNotFoundError`` if absent."""
        ...

    def read_bytes(self, entry: str) -> bytes:
        """Read entry content as bytes. Raises ``FileNotFoundError`` if absent."""
        ...

    def iter_entries(self, *, recursive: bool = False) -> Iterator[str]:
        """Yield entry names (relative to the bundle root).

        When ``recursive=False`` (default), only direct children are
        yielded — both regular files and subdirectories (e.g. ``"scripts"``).
        When ``recursive=True``, descend into subdirectories yielding only
        regular files (no directory entries) using forward-slash separators.
        """
        ...

    def is_file(self, entry: str) -> bool:
        """True if ``entry`` points at a regular file (not a directory).
        Used by readers that filter out subdirs from ``iter_entries()``.
        """
        ...

    def write_text(self, entry: str, content: str, encoding: str = "utf-8") -> None:
        """Write text content to the entry, creating parent dirs as needed.

        Read-only handles MUST raise ``NotImplementedError`` (or a subclass).
        """
        ...

    def write_bytes(self, entry: str, content: bytes) -> None:
        """Write bytes content. Read-only handles raise ``NotImplementedError``."""
        ...

    @property
    def path(self) -> Path | None:
        """Filesystem path when the handle wraps a real directory; ``None``
        otherwise.

        ESCAPE HATCH — prefer the explicit read/write/iter methods. Use this
        only when an external library demands a real ``Path`` (e.g.
        ``shutil.copy``, ``ZipFile``, third-party tooling that takes paths).
        Code paths that need this should gracefully degrade when ``None``.
        """
        ...


# ---------------------------------------------------------------------------
# Filesystem implementation
# ---------------------------------------------------------------------------


class FilesystemBundleHandle:
    """``BundleHandle`` backed by a real filesystem directory.

    Constructed by ``FilesystemSource.load_all`` for each detected bundle
    and passed to the matching reader's ``read(handle)`` method.
    """

    __slots__ = ("_root",)

    def __init__(self, root: Path) -> None:
        self._root = root

    @property
    def name(self) -> str:
        return self._root.name

    def exists(self, entry: str) -> bool:
        return (self._root / entry).exists()

    def read_text(self, entry: str, encoding: str = "utf-8") -> str:
        return (self._root / entry).read_text(encoding=encoding)

    def read_bytes(self, entry: str) -> bytes:
        return (self._root / entry).read_bytes()

    def iter_entries(self, *, recursive: bool = False) -> Iterator[str]:
        if not self._root.is_dir():
            return
        if recursive:
            for child in self._root.rglob("*"):
                if child.is_file():
                    yield child.relative_to(self._root).as_posix()
        else:
            for child in self._root.iterdir():
                yield child.name

    def is_file(self, entry: str) -> bool:
        return (self._root / entry).is_file()

    def write_text(self, entry: str, content: str, encoding: str = "utf-8") -> None:
        target = self._root / entry
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(content, encoding=encoding)

    def write_bytes(self, entry: str, content: bytes) -> None:
        target = self._root / entry
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_bytes(content)

    @property
    def path(self) -> Path | None:
        return self._root


# ---------------------------------------------------------------------------
# In-memory implementation (testing + reader audits)
# ---------------------------------------------------------------------------


class DictBundleHandle:
    """``BundleHandle`` backed by an in-memory ``dict[str, str | bytes]``.

    Use in tests and reader audits to verify that a reader works
    independent of the filesystem. ``path`` returns ``None`` so any
    Path-dependent code path raises a recognisable failure.

    Example:
        >>> handle = DictBundleHandle("my-skill", {
        ...     "SKILL.md": "---\\nname: my-skill\\n---\\nbody",
        ...     "scripts/run.py": "print('hi')",
        ... })
        >>> handle.read_text("SKILL.md")
        '---\\nname: my-skill\\n---\\nbody'
    """

    def __init__(self, name: str, entries: dict[str, str | bytes]) -> None:
        self._name = name
        # Normalise entries: store text as-is, bytes as-is.
        self._entries: dict[str, str | bytes] = dict(entries)

    @property
    def name(self) -> str:
        return self._name

    def exists(self, entry: str) -> bool:
        if entry in self._entries:
            return True
        # Match directory prefixes — e.g. "scripts" exists if any entry
        # starts with "scripts/".
        prefix = entry.rstrip("/") + "/"
        return any(k.startswith(prefix) for k in self._entries)

    def read_text(self, entry: str, encoding: str = "utf-8") -> str:
        v = self._require(entry)
        if isinstance(v, bytes):
            return v.decode(encoding)
        return v

    def read_bytes(self, entry: str) -> bytes:
        v = self._require(entry)
        if isinstance(v, str):
            return v.encode("utf-8")
        return v

    def iter_entries(self, *, recursive: bool = False) -> Iterator[str]:
        if recursive:
            for k in self._entries:
                yield k
            return
        seen: set[str] = set()
        for k in self._entries:
            top = k.split("/", 1)[0]
            if top not in seen:
                seen.add(top)
                yield top

    def is_file(self, entry: str) -> bool:
        return entry in self._entries

    def write_text(self, entry: str, content: str, encoding: str = "utf-8") -> None:
        self._entries[entry] = content

    def write_bytes(self, entry: str, content: bytes) -> None:
        self._entries[entry] = content

    @property
    def path(self) -> Path | None:
        return None

    def _require(self, entry: str) -> str | bytes:
        try:
            return self._entries[entry]
        except KeyError as e:
            raise FileNotFoundError(
                f"DictBundleHandle({self._name!r}) has no entry {entry!r}"
            ) from e


__all__ = [
    "BundleHandle",
    "FilesystemBundleHandle",
    "DictBundleHandle",
]
